crates/containerd-shim-wasm/src/sandbox/containerd.rs

```diff
@@ -0,0 +1,148 @@
+#![cfg(unix)]
+
+use std::path::Path;
+
+use containerd_client;
+use containerd_client::services::v1::containers_client::ContainersClient;
+use containerd_client::services::v1::content_client::ContentClient;
+use containerd_client::services::v1::images_client::ImagesClient;
+use containerd_client::services::v1::{GetContainerRequest, GetImageRequest, ReadContentRequest};
+use containerd_client::tonic::transport::Channel;
+use containerd_client::{tonic, with_namespace};
+use futures::TryStreamExt;
+use oci_spec::image::{ImageManifest, Os};
+use tokio::runtime::Runtime;
+use tonic::Request;
+
+use crate::sandbox::error::{Error as ShimError, Result};
+use crate::sandbox::oci::{self, WasmLayer};
+
+pub struct Client {
+    inner: Channel,
+    rt: Runtime,
+    namespace: String,
+}
+
+// sync wrapper implementation from https://tokio.rs/tokio/topics/bridging
+impl Client {
+    // wrapper around connection that will establish a connection and create a client
+    pub fn connect(address: impl AsRef<Path>, namespace: impl ToString) -> Result<Client> {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()?;
+
+        let inner = rt
+            .block_on(containerd_client::connect(address))
+            .map_err(|err| ShimError::Others(err.to_string()))?;
+
+        Ok(Client {
+            inner,
+            rt,
+            namespace: namespace.to_string(),
+        })
+    }
+
+    // wrapper around read that will read the entire content file
+    pub fn read_content(&self, digest: impl ToString) -> Result<Vec<u8>> {
+        self.rt.block_on(async {
+            let req = ReadContentRequest {
+                digest: digest.to_string(),
+                ..Default::default()
+            };
+            let req = with_namespace!(req, self.namespace);
+            ContentClient::new(self.inner.clone())
+                .read(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?
+                .into_inner()
+                .map_ok(|msg| msg.data)
+                .try_concat()
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))
+        })
+    }
+
+    pub fn get_image_content_sha(&self, image_name: impl ToString) -> Result<String> {
+        self.rt.block_on(async {
+            let name = image_name.to_string();
+            let req = GetImageRequest { name };
+            let req = with_namespace!(req, self.namespace);
+            let digest = ImagesClient::new(self.inner.clone())
+                .get(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?
+                .into_inner()
+                .image
+                .ok_or(ShimError::Others(format!(
```

Comment:
Can we switch this (and the others) `ok_or` with `ok_or_else` to avoid the `format!` when it's not needed ?